### PR TITLE
fix(publish): Raise MQTTPublishError on negative PUBACK/PUBREC reason codes.

### LINKED ISSUE #65: MQTT 5 publish() silently succeeds when PUBACK or PUBREC rejects the message
## Summary

`MQTTClient.publish()` returns normally when an MQTT 5 broker rejects a QoS 1 or QoS 2 PUBLISH with a negative acknowledgement reason code.

For example, Mosquitto returns `0x87 (Not authorized)` for an ACL-denied publication:

```text
Received PubAck(packet_id=1, reason_code=135, properties=None)
```

The message was not accepted by the broker, but `await client.publish(...)` completes without an exception.

## Current behavior

For QoS 1, `_handle_puback()` resolves the publish future successfully without checking `PubAck.reason_code`.

For QoS 2, `_handle_pubrec()` also ignores the reason code and sends PUBREL even when PUBREC contains `0x87`. MQTT 5 only permits PUBREL after a PUBREC reason code below `0x80`.

The public `MQTTClient.publish()` method discards the acknowledgement returned by `MQTTProtocol.publish()`.

## Expected behavior

- PUBACK/PUBREC reason codes below `0x80`, including `0x00` and `0x10 (No matching subscribers)`, complete successfully.
- A PUBACK or PUBREC reason code of `0x80` or greater raises a public `MQTTPublishError`.
- The exception exposes at least `reason_code`; exposing the optional Reason String would also be useful.
- A negative PUBREC completes and removes the QoS 2 flight, releases its packet identifier, and does not send PUBREL.
- The MQTT connection remains usable after the rejected operation.
- MQTT 3.1.1 behavior remains unchanged because PUBACK and PUBREC do not carry reason codes in that protocol version.

`MQTTPublishError` would be consistent with the existing `MQTTSubscribeError`.

## Protocol references

- [MQTT 5.0 section 2.4](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901029): reason codes `0x80` and greater indicate failure.
- [MQTT 5.0 section 3.4.2.1](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901125): PUBACK `0x87` means the PUBLISH is not authorized.
- [MQTT 5.0 section 3.5.2.1](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901131): PUBREC has the same publish-rejection reason codes.
- [MQTT 5.0 section 4.3.3](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901235): PUBREL is sent only after a PUBREC reason code below `0x80`.
- [MQTT 5.0 section 4.4](https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc3901236): a negative PUBACK/PUBREC acknowledges the packet for retry purposes, but does not make the publication successful.
- [MQTT 3.1.1 section 3.3.5](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718043): when a PUBLISH is not authorized, the server must either send a positive acknowledgement or close the connection because the protocol has no negative publish acknowledgement.

## Reproduction result

Reproduced with Mosquitto 2.1.2 and a read-only ACL:

```text
MQTT 5 QoS 1:
Received PubAck(packet_id=1, reason_code=135, properties=None)
publish() returned normally

MQTT 5 QoS 2:
Received PubRec(packet_id=1, reason_code=135, properties=None)
QoS 2 PUBREC received, sent PUBREL
publish() returned normally
```

Mosquitto logs `Denied PUBLISH` for both messages.

## Suggested tests

- QoS 1 negative PUBACK raises and releases the packet identifier.
- QoS 1 `0x10` completes successfully.
- QoS 2 negative PUBREC raises, releases the packet identifier, and sends no PUBREL.
- Optional Reason String is preserved in the exception.
- MQTT 3.1.1 ACK behavior is unchanged.

### PR DESCRIPTION
**Pull request by lubaskinc0de**
## Summary
Raise MQTTPublishError on negative PUBACK/PUBREC reason codes. Fixes #65 
Note: i also added .vscode to gitignore cause it's a common thing created by vscode for many devs

BREAKING CHANGE: On MQTT 5.0, a QoS 1/2 publish rejected by the broker (PUBACK/PUBREC reason code >= 0x80) now raises MQTTPublishError instead of returning normally. MQTT 3.1.1 and QoS 0 are unchanged.

## Validation

- [x] Tests were added or updated when behavior changed
- [x] Documentation was updated when the public API changed
- [x] The PR title follows Conventional Commits, for example `feat(client): add reconnect timeout`

### CODE REVIEW

File: tests/test_brokers/_base.py [hunk 1]
```diff
@@ -33,6 +35,8 @@ class BrokerTestBase(abc.ABC):
     port: ClassVar[int] = 1883
     version: ClassVar[Literal["3.1.1", "5.0"]] = "3.1.1"
     supports_persistent_sessions: ClassVar[bool] = True
+    has_publish_deny_acl: ClassVar[bool] = False
```
Comment (borisalekseev) [inline tests/test_brokers/_base.py]: Just write test in Mosquitto mqtt 5 test class

File: tests/test_brokers/_base.py [hunk 2]
```diff
@@ -33,6 +35,8 @@ class BrokerTestBase(abc.ABC):
...
+    denied_topic: ClassVar[str] = "zmqtt/e2e/denied"
```
Comment (borisalekseev) [inline tests/test_brokers/_base.py]: move to constant in mosquitto mqtt 5 tests

File: tests/test_protocol.py
Comment (borisalekseev) [inline tests/test_protocol.py:1]: Drop all these tests, e2e covers scenario and fix behaviour well

**Review by borisalekseev — CHANGES_REQUESTED** [review]
Thanks! Wait for test fixes

File: tests/test_protocol.py
Comment (lubaskinc0de) [reply tests/test_protocol.py:1]: i think that e2e cannot truly cover other reason codes (it only cover 0x87) + this expected behaviour points:
- The exception exposes at least reason_code; exposing the optional Reason String would also be useful. (cause mosquito do not returning reason string)
- A negative PUBREC completes and removes the QoS 2 flight, releases its packet identifier, and does not send PUBREL.(we cant check that it doesn't send pubrel and releases packet id)

isn't it?

Comment (lubaskinc0de) [reply tests/test_protocol.py:1]: also it cant cover this corner-case: test_puback_failure_with_unknown_error_code_sets_reason_name_none

Comment (borisalekseev) [reply tests/test_protocol.py:1]: First one tested OK, we covered the branch that checks if the code is greater than 0x80 and raises an error. Agree with the second one, we can leave it as is. One test case for the fact that PUBREL is not sent will be sufficient here. 100% coverage is not the goal.

File: tests/test_brokers/_base.py (hunk 1 above)
Comment (lubaskinc0de) [reply tests/test_brokers/_base.py]: fixed

File: tests/test_brokers/_base.py (hunk 2 above)
Comment (lubaskinc0de) [reply tests/test_brokers/_base.py]: fixed

File: tests/test_protocol.py
Comment (lubaskinc0de) [reply tests/test_protocol.py:1]: done

**Review by borisalekseev — APPROVED** [review]
